src/stan/services/pathfinder/single.hpp

```diff
@@ -0,0 +1,1201 @@
+    debug_stream << "Alpha: \n" << alpha.format(CommaInitFmt) << "\n";
+    debug_stream << "Qk: \n" << Qk.format(CommaInitFmt) << "\n";
+    debug_stream << "L_approx: \n" << L_approx.format(CommaInitFmt) << "\n";
+    debug_stream << "logdetcholHk: \n" << logdetcholHk << "\n";
+    debug_stream << "Mkbar: \n" << Mkbar.format(CommaInitFmt) << "\n";
+    debug_stream << "Decomp Wkbar: \n" << Wkbart.format(CommaInitFmt) << "\n";
+    debug_stream << "x_center: \n" << x_center.format(CommaInitFmt) << "\n";
+    debug_stream << "NinvRST: " << ninvRST.format(CommaInitFmt) << "\n";
+    debug_stream << "ninvRSTg: \n" << ninvRSTg.format(CommaInitFmt) << "\n";
+    debug_stream << "Rkbar: " << Rkbar.format(CommaInitFmt) << "\n";
+    logger.info(debug_stream);
+  }
+}
+}  // namespace debug
+template <typename T1, typename T2>
+inline auto crossprod(T1&& x, T2&& y) {
+  return x.transpose() * y;
+}
+
+template <typename T1>
+inline Eigen::MatrixXd crossprod(T1&& x) {
+  return Eigen::MatrixXd(x.cols(), x.cols())
+      .setZero()
+      .selfadjointView<Eigen::Lower>()
+      .rankUpdate(x.adjoint());
+}
+
+template <typename T1, typename T2>
+inline auto tcrossprod(T1&& x, T2&& y) {
+  return x * y.transpose();
+}
+
+template <typename T1>
+inline Eigen::MatrixXd tcrossprod(T1&& x) {
+  return Eigen::MatrixXd(x.rows(), x.rows())
+      .setZero()
+      .selfadjointView<Eigen::Lower>()
+      .rankUpdate(x);
+}
+
+/**
+ * Perform a `Matrix * vector.asDiagonal()` multiplication with the matrix
+ * represented as an `std::vector` of rows.
+ * @tparam EigVec1 A type inheriting from `Eigen::DenseBase` with compile time
+ *  rows or columns equal to 1
+ * @tparam EigVec2 A type inheriting from `Eigen::DenseBase` with compile time
+ *  rows or columns equal to 1
+ * @param y_buff A standard vector with elements representing the rows of
+ *  a matrix.
+ * @param alpha An eigen vector representing the diagonals of a matrix.
+ * @return Returns the same result as if we called `Matrix *
+ * vector.asDiagonal()`
+ */
+template <typename EigVec1, typename EigVec2>
+inline Eigen::MatrixXd std_vec_matrix_times_diagonal(
+    const std::vector<EigVec1>& y_buff, const EigVec2& alpha) {
+  Eigen::MatrixXd ret(y_buff.size(), alpha.size());
+  for (Eigen::Index i = 0; i < y_buff.size(); ++i) {
+    ret.row(i) = y_buff[i].array() * alpha.array();
+  }
+  return ret;
+}
+
+/**
+ * Perform a `Matrix.transpose() * vector` multiplication with the matrix
+ * represented as an `std::vector` of rows.
+ * @tparam EigVec1 A type inheriting from `Eigen::DenseBase` with compile time
+ *  rows or columns equal to 1
+ * @tparam EigVec2 A type inheriting from `Eigen::DenseBase` with compile time
+ *  rows or columns equal to 1
+ * @param y_buff A standard vector with elements representing the rows of
+ *  a matrix.
+ * @param alpha An eigen vector.
+ * @return Returns the same result as if we called `Matrix.transpose() * vector`
```

Comment:
The intent of this is to show that this function takes an `std::vector<Eigen::VectorXd>` and `Eigen::VectorXd` and gives back a result as if the `std::vector<Eigen::VectorXd>` was an `Eigen::MatrixXd` and we did `Eigen::MatrixXd.transpose() * Eigen::VectorXd`. What is the better way to write document this? Would it help if I put `EIgen::` before `Matrix` and `vector`?

Reply:
I think a member function should be written as `Eigen::Matrix::transpose`?  At least that's the syntax used in the C++ doc.  I was thinking that `Matrix` was an object being transposed because of the period.